src/config.rs

```diff
@@ -108,10 +114,111 @@ pub fn clear_thread_local_config_option(key: &str) -> Result<()> {
     Ok(())
 }
 
+type CallbackType = dyn FnMut(CplErrType, i32, &str) + 'static + Send;
+type PinnedCallback = Pin<Box<Box<CallbackType>>>;
+
+/// Static variable that holds the current error callback function
+static ERROR_CALLBACK: Lazy<Mutex<Option<PinnedCallback>>> = Lazy::new(Default::default);
+
+/// Set a custom error handler for GDAL.
+/// Could be overwritten by setting a thread-local error handler.
+///
+/// Note:
+/// Stores the callback in the static variable [`ERROR_CALLBACK_THREAD_SAFE`].
+/// Internally, it passes a pointer to the callback to GDAL as `pUserData`.
+///
+pub fn set_error_handler<F>(callback: F)
+where
+    F: FnMut(CplErrType, i32, &str) + 'static + Send,
+{
+    unsafe extern "C" fn error_handler(
+        error_type: CPLErr::Type,
+        error_num: CPLErrorNum,
+        error_msg_ptr: *const c_char,
+    ) {
+        let error_msg = _string(error_msg_ptr);
+        let error_type: CplErrType = error_type.into();
+
+        // reconstruct callback from user data pointer
+        let callback_raw = CPLGetErrorHandlerUserData();
+        let callback: &mut Box<CallbackType> = &mut *(callback_raw as *mut Box<_>);
+
+        callback(error_type, error_num as i32, &error_msg);
+    }
+
+    // pin memory location of callback for sending its pointer to GDAL
+    let mut callback: PinnedCallback = Box::pin(Box::new(callback));
+
+    let callback_ref: &mut Box<CallbackType> = callback.as_mut().get_mut();
+    unsafe {
+        gdal_sys::CPLSetErrorHandlerEx(Some(error_handler), callback_ref as *mut _ as *mut c_void);
+    };
+
+    let mut callback_lock = match ERROR_CALLBACK.lock() {
```

Comment:
I think there may be a race condition here.  Suppose two threads A, B race to set a error handler.  Then, it is possible A sets the handler, then B sets, thus replacing it (from GDAL's perspective).  However, during the subsequent Mutex lock, the order could get reversed, and the mutex ends up storing the wrong handler; worse yet, the correct handler has now been dropped, and could cause a crash when accessed.

Reply:
I was able to modify the test case you had, to show how it could trigger:

```rust
#[test]
fn error_handler() {
    use std::thread;
    // Two racing threads trying to set error handlers
    // First one
    thread::spawn(move || {
        loop {
            set_error_handler(move |a, b, c| {
            });
        }
    });

    // Second one
    thread::spawn(move || {
        loop {
            set_error_handler(move |a, b, c| {
            });
        }
    });


    // A thread that makes a lot of mistakes
    thread::spawn(move || {
        loop {
            unsafe {
                let msg = CString::new("foo".as_bytes()).unwrap();
                gdal_sys::CPLError(CPLErr::CE_Failure, 42, msg.as_ptr());
            };

            unsafe {
                let msg = CString::new("bar".as_bytes()).unwrap();
                gdal_sys::CPLError(std::mem::transmute(CplErrType::Warning), 1, msg.as_ptr());
            };
        }
    });

    // Wait for the segfault to happen.  Happens within a few seconds typically.
    loop {}
}
```

Reply:
provided a fix and included the test (kind of modified due to the infinite loops)

bce691d